### PR TITLE
add integration tests to coverage

### PR DESCRIPTION
**Pull request by ailisp**

### CODE REVIEW

**Comment by ailisp** [comment]
@nearmax result turns out to be good - 90% now, but coverage time increase from 20min to 30min, this seems slow down our ci. We might consider don't run coverage on every build or allow ci to pass after test_cargo/test_nearlib is success, and allow coverage to be report later.

File: scripts/coverage.sh [hunk 1]
```diff
@@ -5,16 +5,15 @@ if [[ -z "${GITLAB_CI}" ]]; then
   exit 1
 fi
 
-for file in `find target/debug/deps/ \
-  ! -name 'test*' \
+ for file in `find target/debug/deps/ \
```
Comment (nearmax) [inline scripts/coverage.sh]: Can we exclude top-level integration tests in `nearcore/tests`?

**Comment by ailisp** [comment]
Hmm result looks weird, it still takes 30 minutes, and after exclude top level integration tests time reduce to 79.% again

**Comment by ailisp** [comment]
Good news is https://codecov.io/gh/nearprotocol/nearcore/pull/1761/src/runtime/near-vm-logic/src/logic.rs does fully covered now.

I manually run kcov on top level integration tests, they're all very fast. So the 10 minute increase comes from some crate level integration test. Let me see identify and exclude them. Shall we include the top level integration tests back @nearmax

**Comment by ailisp** [comment]
I measures how long each `kcov` takes on all our tests (in second). In ci every time takes ~2x time, if anyone would suggest some test to exclude or include in cov Please reply here. Current status: all top level integration tests are included.
````
('target/debug/deps/test_regression-c807fdb64d8c9aa4', 227.792)
('target/debug/deps/chunks_management-b8cb95471fa9ae7e', 114.27000000000001)
('target/debug/deps/run_nodes-97f7f12acb5b1352', 66.486)
('target/debug/deps/stake_nodes-c67a256c1e06c0aa', 62.278)
('target/debug/deps/sync_nodes-a522ebf590a58b0f', 55.61)
('target/debug/deps/sync_state_nodes-8cb8f05603ae51f4', 53.286)
('target/debug/deps/test_cases_runtime-7f3e4cae3d4f8cd8', 41.359)
('target/debug/deps/rpc_nodes-54297109985b4e55', 40.782)
('target/debug/deps/test_rs_contract-846d0a7dee695f99', 37.394)
('target/debug/deps/track_shards-b1ad5eeae0f449e7', 28.888)
('target/debug/deps/announce_account-eaacc12493916a18', 27.595)
('target/debug/deps/bug_repros-a8947500f7476dd8', 24.734)
('target/debug/deps/test_errors-6a15733b2d68c1c8', 22.4)
('target/debug/deps/http_query-4e3f185bde781561', 21.355)
('target/debug/deps/routing-b833601755f5839e', 19.211)
('target/debug/deps/rpc_transactions-41a573cd09b87292', 17.277)
('target/debug/deps/test_evil_contracts-53ff5397b55f8e9c', 15.791)
('target/debug/deps/rpc_query-4fba81421b56c09a', 15.762)
('target/debug/deps/process_blocks-604b2c7ca7ebe452', 14.027)
('target/debug/deps/catching_up-74ed03fcb2bf65ad', 13.567)
('target/debug/deps/test_async_calls-56194c7bc40337a7', 12.952)
('target/debug/deps/cross_shard_tx-e05b6f8c263af0e3', 12.459)
('target/debug/deps/near_store-f5fe0b1d8f162c05', 11.993)
('target/debug/deps/near_chain-4ee505f9fb1f7e03', 10.979)
('target/debug/deps/challenges-e49e80d0f8b48d65', 10.815)
('target/debug/deps/query_client-b35fe46e32c1139e', 9.753)
('target/debug/deps/testlib-b16668d827543c0d', 9.549)
('target/debug/deps/node_runtime-db868525e647a42f', 9.164)
('target/debug/deps/finality-5abbf3117be77519', 8.465)
('target/debug/deps/peer_handshake-3604bc61467d09ea', 7.463)
('target/debug/deps/simple_chain-7ca57595355bf1d5', 6.193)
('target/debug/deps/near_client-a980fca0886c07d9', 5.731)
('target/debug/deps/near_primitives-26ac3aee406a0a66', 5.56)
('target/debug/deps/challenges-64faa4cfb87636cf', 5.534)
('target/debug/deps/near_network-4917fd7e0a97d9ea', 5.255)
('target/debug/deps/genesis_csv_to_json-3e5d701602cdfdbf', 4.383)
('target/debug/deps/sync_chain-c455bdcd757d891b', 3.668)
('target/debug/deps/test_registers-dbff44b06ca94ed7', 3.61)
('target/debug/deps/test_ts_contract-2fac3a706fc6d494', 3.526)
('target/debug/deps/test_invalid_contracts-b06c5383b2139c69', 3.367)
('target/debug/deps/test_error_cases-ec18a9f77f704f29', 3.286)
('target/debug/deps/near_epoch_manager-d786cc76c6d7e329', 2.937)
('target/debug/deps/near_jsonrpc_client-ba7f5f1a1873d570', 2.597)
('target/debug/deps/near_vm_runner-40bdd755869883d3', 1.593)
('target/debug/deps/test_iterators-a531f7789a9c9f45', 1.592)
('target/debug/deps/near_pool-624245e5393a3167', 0.999)
('target/debug/deps/test_promises-eaf536c90c5cec41', 0.955)
('target/debug/deps/test_cases_testnet_rpc-5bc3eff31b0ddc6d', 0.92)
('target/debug/deps/near_crypto-7ebd0a75c27c3ff9', 0.888)
('target/debug/deps/runtime_params_estimator-aea0f3b53dd02aba', 0.861)
('target/debug/deps/test_view_method-5fddd84cbdcc6361', 0.629)
('target/debug/deps/test_miscs-4bb6e0c2d7581c36', 0.61)
('target/debug/deps/test_storage_read_write-de1d23d4c114c6d4', 0.581)
('target/debug/deps/genesis_populate-cc1f9c499d454aca', 0.529)
('target/debug/deps/test_context-e279a369847955f1', 0.505)
('target/debug/deps/test_storage_usage-d84604e2c0c53dd6', 0.49)
('target/debug/deps/fixtures-6f655a37225cfd20', 0.479)
('target/debug/deps/near_jsonrpc-fdee87f8216a457e', 0.425)
('target/debug/deps/near_metrics-db9809454319181c', 0.37)
('target/debug/deps/near_vm_errors-91abcde469cc966e', 0.351)
('target/debug/deps/near_vm_logic-de7ca5177440c6a9', 0.339)
('target/debug/deps/finality-fbc732e8cefd145d', 0.324)
('target/debug/deps/async_utils-421deed071413686', 0.316)
('target/debug/deps/test_tps_regression-342236be7cc31e11', 0.307)
('target/debug/deps/keypair_generator-bbc3f1b4fc1e5b1e', 0.295)
('target/debug/deps/state_viewer-2a68875f58984480', 0.294)
('target/debug/deps/near_runtime_fees-d748f723d185392d', 0.289)
('target/debug/deps/helpers-010d6cbfebe234b6', 0.289)
('target/debug/deps/test_simple-1a1f3c29c17df561', 0.286)
('target/debug/deps/run_nodes-b4be2b7030bf0fd4', 0.285)
('target/debug/deps/test_catchup-9d3d726de213434a', 0.282)
('target/debug/deps/utils-dbaf2bcb12a8c4df', 0.28)
('target/debug/deps/near_chunks-36bece3cc496c365', 0.279)
('target/debug/deps/test_rejoin-f9b94713c72f181d', 0.278)
('target/debug/deps/near_vm_runner_standalone-867d48bc99a5f32d', 0.272)
('target/debug/deps/vm_logic_builder-dc0b99f7f7024a23', 0.271)
('target/debug/deps/genesis_populate-96ef84a2c1950746', 0.269)
('target/debug/deps/near_telemetry-607db0602e45a870', 0.268)
('target/debug/deps/runtime_params_estimator-2bd7e7d926db516c', 0.26)
```

**Comment by nearmax** [comment]
> Good news is https://codecov.io/gh/nearprotocol/nearcore/pull/1761/src/runtime/near-vm-logic/src/logic.rs does fully covered now.
> 
> I manually run kcov on top level integration tests, they're all very fast. So the 10 minute increase comes from some crate level integration test. Let me see identify and exclude them. Shall we include the top level integration tests back @nearmax

My concern is not that top-level integration tests are slow, but that they misrepresent the coverage. The top-level tests trigger almost the entire infrastructure. So I suspect if we leave out a single top-level integration test (e.g. one that does function call) our coverage is still going to be pretty high.

I think it is a terminology issue, in Rust they call "integration test" everything that is in `tests` folder and not inside a module. However, classically integration test are checking *integration* between components. So according to classical definition tests in `near-vm-logic/tests` are not integration tests. I think it is a good idea to exclude classical integration tests (the ones in `nearcore/tests`, but still include crate-level tests, like `near-vm-logic/tests`. In fact as we get our code more and more cleaner and encapsulated we are going to have fewer unit tests and more crate-level tests, because with well-encapsulated and mockable code there are fewer needs to tests internals (Rust unit tests have access to private fields/methods).

**Comment by nearmax** [comment]
> Good news is https://codecov.io/gh/nearprotocol/nearcore/pull/1761/src/runtime/near-vm-logic/src/logic.rs does fully covered now.

Cool, but it is still weird that it has <100% coverage, while all of the executable lines of code are green.

**Comment by ailisp** [comment]
Got it. Thanks for explaining the integration test and good to know top
level real integration test gives a false prediction on coverage. Why is
not 100% might be limitation of how codecov count rust lines.

**Pull request merged**